### PR TITLE
doc: usb: explain interface number and endpoint address assignment

### PR DESCRIPTION
**Pull request by jfischer-no**
This also applies to the new experimental device stack, but documentation for it has not yet begun.

### CODE REVIEW

File: doc/connectivity/usb/device/usb_device.rst [hunk 1]
```diff
@@ -447,6 +447,70 @@ before transmitting any data.
...
+The actual number depends on the device controller used. Not all controllers
+support the maximum number of endpoints and all endpoint types. For example, a
+device controller might support one IN and one OUT isochronous endpoint, but
+only for endpoint number 8, resulting in endpoint addresses 0x88 and 0x08.
+Also, one controller may be able to have IN/OUT endpoints on the same endpoint
+number, interrupt IN endpoint 0x81 and bulk OUT endpoint 0x01, while the other
+may only be able to handle one endpoint per endpoint number. Information about
+the number of interfaces, interface associations, endpoint types, and addresses
+is provided to the host by the interface, interface specifiec, and endpoint
+descriptors.
```
Comment (tmon-nordic) [inline doc/connectivity/usb/device/usb_device.rst:478]: A common misinterpretation I have observed is that people tend to consider 0x81 and 0x01 as a single endpoint working in both directions. However, in reality, any non-control endpoint is always unidirectional, meaning that 0x81 and 0x01 are completely independent endpoints just like 0x82 from 0x01 is. Not sure how to concisely write it.

Also while specification allows non-zero control endpoints, I don't think Zephyr supports it.

File: doc/connectivity/usb/device/usb_device.rst [hunk 2]
```diff
@@ -447,6 +447,70 @@ before transmitting any data.
...
+meaning it cannot be used with different device controllers and different
+device configurations in combination with other functions. This may also be
+because the driver is designed for a specific hardware and is not intended to
+be used with a clone of this specific hardware. On the contrary, if the driver
+is generic in nature and should work with different hardware variants, then it
+must not use hardcoded interface numbers and endpoint addresses.
+It is not possible to disable endpoint reassignment in Zephyr RTOS, which may
+prevent you from implementing a hardware-clone firmware. Instead, if possible,
+the host driver implementation should be fixed to use values from the interface
+and endpoint descriptor.
```
Comment (tmon-nordic) [inline doc/connectivity/usb/device/usb_device.rst:512]: I wish it was possible to do endpoint/interface fixup at compile/linking time, but that would require some serious build system development.

**Review by tmon-nordic — DISMISSED** [review]
Correct the typo and it's good. Added two side notes, but the point of this PR is documenting current implementation and not necessarily what is on the wishlist.

File: doc/connectivity/usb/device/usb_device.rst (hunk 2 above)
Comment (jfischer-no) [reply doc/connectivity/usb/device/usb_device.rst:512]: It is not that expensive to do this at runtime, and compared to what people are willing to sacrifice resources for, it is not worth thinking about. Our behavior is very similar to the Linux gadget drivers. And it provides the feature that you can be very flexible with function configuration, even in multiple configurations, at runtime. So one can shutdown the stack and reconfigure it to have a different function and enable it again. We had a feature request for that.

File: doc/connectivity/usb/device/usb_device.rst (hunk 1 above)
Comment (jfischer-no) [reply doc/connectivity/usb/device/usb_device.rst:478]: Maybe some controller documentation is to blame, and our binding description does not make it better. e.g. from STM32 RM0431 "Device bidirectional endpoints (including EP0)" ended up in dts/bindings/usb/usb-ep.yaml


```
      Number of bi-directional endpoints supported by hardware
      (including EP0)
```

**Review by tmon-nordic — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**